ipalib/rpc.py

```diff
@@ -274,102 +274,171 @@ def xml_dumps(params, version, methodname=None, methodresponse=False,
+            tuple: self._enc_list,
+            dict: self._enc_dict,
+        })
+        # int, long
+        for t in six.integer_types:
+            self[t] = _identity
+
+    def __missing__(self, typ):
+        # walk MRO to find best match
+        for c in typ.__mro__:
+            if c in self:
+                self[typ] = self[c]
+                return self[c]
+        # use issubclass to check for registered ABCs
+        for c in self:
+            if issubclass(typ, c):
+                self[typ] = self[c]
+                return self[c]
+        raise TypeError(typ)
+
+    def convert(self, obj, _identity=_identity):
+        # obj.__class__ is twice as fast as type(obj)
+        func = self[obj.__class__]
+        return obj if func is _identity else func(obj)
+
+    def _enc_datetime(self, val):
+        cap = self._cap_datetime
+        if cap is None:
+            cap = capabilities.client_has_capability(self.version,
+                                                     'datetime_values')
+            self._cap_datetime = cap
+        if cap:
             return {'__datetime__': val.strftime(LDAP_GENERALIZED_TIME_FORMAT)}
         else:
             return val.strftime(LDAP_GENERALIZED_TIME_FORMAT)
-    elif isinstance(val, DNSName):
-        if capabilities.client_has_capability(version, 'dns_name_values'):
+
+    def _enc_dnsname(self, val):
+        cap = self._cap_dnsname
+        if cap is None:
+            cap = capabilities.client_has_capability(self.version,
+                                                     'dns_name_values')
+            self._cap_dnsname = cap
+        if cap:
             return {'__dns_name__': unicode(val)}
         else:
             return unicode(val)
-    elif isinstance(val, Principal):
-        return unicode(val)
+
+    def _enc_bytes(self, val):
+        encoded = base64.b64encode(val)
+        if not six.PY2:
+            encoded = encoded.decode('ascii')
+        return {'__base64__': encoded}
+
+    def _enc_list(self, val, _identity=_identity):
+        result = []
+        append = result.append
+        for v in val:
+            func = self[v.__class__]
+            append(v if func is _identity else func(v))
+        return result
+
+    def _enc_dict(self, val, _identity=_identity, _iteritems=six.iteritems):
+        result = {}
+        for k, v in _iteritems(val):
+            func = self[v.__class__]
+            result[k] = v if func is _identity else func(v)
+        return result
+
+
+def json_encode_binary(val, version, pretty_print=False):
+    """Serialize a Python object structure to JSON
+
+    :param object val: Python object structure
+    :param str version: client version
+    :param bool pretty_print: indent and sort JSON (warning: slow!)
+    :return: text
+    :note: pretty printing triggers a slow path in Python's JSON module. Only
+           use pretty_print in debug mode.
+    """
+    result = _JSONPrimer(version).convert(val)
```

Comment:
Wouldn't it be a little bit faster if we kept the type - encoder mapping in a global object initialized once rather than reconstructing it in every `json_encode_binary()` call?